### ISSUE TITLE
Visualizing Concurrency in Go  divan's blog

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#3026]
Visualizing Concurrency in Go · divan's blog<br>
<br>
https://ift.tt/2CmhnoE<br>
<br>
Gopher. Public speaker. Available for hire.
			Follow me at @idanyliuk<br>
<br>
<div><section>
<p><em>If you prefer video over blog posts, here is my talk on this at GopherCon 2016:</em>
<a href="https://www.youtube.com/watch?v=KyuFeiG3Y60">https://www.youtube.com/watch?v=KyuFeiG3Y60</a></p>
<p>One of the strongest sides of Go programming language is a built-in concurrency based on <a href="https://en.wikipedia.org/wiki/Communicating_sequential_processes">Tony Hoare’s CSP</a> paper. Go is designed with concurrency in mind and allows us to build complex concurrent pipelines. But have you ever wondered - how various concurrency patterns look like?</p>
<p>Of course, you have. We’re all thinking mostly by visualization in one form or another. If I ask you something involving “numbers from 1 to 100” you will have your own image of the series in your head, even without realizing it. For example, I imagine it as a line going from me with numbers from 1 to 20, then it turns 90 degrees to the right and continues to the 1000+. I recall from very young period of my life that in our kindergarten there were numbers in a cloakroom, written along the wall, and number 20 was exactly at the corner. You probably have your own image of numbers. Another common example is the visual representation of the full year with four seasons - some people see it as a box, other - as a circle.</p>
<p>Anyway, I want to show you my attempt to visualize common concurrency patterns using Go and WebGL. It more or less represents the way I see concurrent programs in my head. Would be pretty interesting to hear how much it differs from images in your head. I especially would love to see how Rob Pike or Sameer Ajmani imagine concurrency. I bet I’d be quite interesting.</p>
<p>So, let’s start with the very basic “Hello, Concurrent World” example, to get ourselves familiar with the concept of my attempt.</p>
<h3>Hello, Concurrent world</h3>
<p>The code is quite simple - single channel, single goroutine, one write, one read.</p>
<pre><code><span>package</span> main

<span>func</span> main() {
    
    ch := <span>make</span>(<span>chan</span> <span>int</span>)

    
    <span>go</span> <span>func</span>() {
        
        ch &lt;- <span>42</span>
    }()
    
    &lt;-ch
}</code></pre>
<p><a href="https://divan.dev/demos/hello/">Go to interactive WebGL animation</a>
<img alt="Hello, World" src="https://divan.dev/demos/gifs/hello.gif"></p>
<p>Here the blue lines represent goroutines running down through time. Thin blue lines connecting ‘main’ and ‘go #19’ are marks for start and stop goroutine, revealing parent-children relation and, finally, red arrow shows us the send/recv action. While it’s actually two separate actions, I try to animate as a single event “send from A to B”. The “#19” in the goroutine name is the actual goroutine internal ID, obtained from runtime with a trick suggested by Scott Mansfield in <a href="http://blog.sgmansfield.com/2015/12/goroutine-ids/">“Goroutine IDs”</a> article.</p>
<h3>Timers</h3>
<p>In fact, you can build a simple timer with this approach - create a channel, start goroutine which writes to this channel after given duration and returns this channel to the caller of your func. The caller then blocks on reading from the channel for the exact amount of time. Let’s run such timer 24 times and try to visualize it.</p>
<pre><code><span>package</span> main

<span>import</span> <span>"time"</span>

<span>func</span> timer(d time.Duration) &lt;-<span>chan</span> <span>int</span> {
    c := <span>make</span>(<span>chan</span> <span>int</span>)
    <span>go</span> <span>func</span>() {
        time.Sleep(d)
        c &lt;- <span>1</span>
    }()
    <span>return</span> c
}

<span>func</span> main() {
    <span>for</span> i := <span>0</span>; i &lt; <span>24</span>; i++ {
        c := timer(<span>1</span> * time.Second)
        &lt;-c
    }
}</code></pre>
<p><a href="https://divan.dev/demos/timers/">Go to interactive WebGL animation</a>
<img alt="Recurrent Timers" src="https://divan.dev/demos/gifs/timers.gif"></p>
<p>Pretty neat, right? Let’s move on.</p>
<h3>Ping-pong</h3>
<p>This nice concurrency example was found in a great talk by googler Sameer Ajmani <a href="https://talks.golang.org/2013/advconc.slide#1">“Advanced Go Concurrency Patterns”</a>. Of course, this pattern isn’t very advanced, but for those who only get themselves familiar with Go concurrency it may look quite fresh and interesting.</p>
<p>Here we have a channel as a table of the ping-pong game. The ball is an integer variable, and two goroutines-players that ‘hit’ the ball, increasing its value (hits counter).</p>
<pre><code><span>package</span> main

<span>import</span> <span>"time"</span>

<span>func</span> main() {
    <span>var</span> Ball <span>int</span>
    table := <span>make</span>(<span>chan</span> <span>int</span>)
    <span>go</span> player(table)
    <span>go</span> player(table)

    table &lt;- Ball
    time.Sleep(<span>1</span> * time.Second)
    &lt;-table
}

<span>func</span> player(table <span>chan</span> <span>int</span>) {
    <span>for</span> {
        ball := &lt;-table
        ball++
        time.Sleep(<span>100</span> * time.Millisecond)
        table &lt;- ball
    }
}</code></pre>
<p><a href="https://divan.dev/demos/pingpong/">Go to interactive WebGL animation</a>
<img alt="Ping-Pong" src="https://divan.dev/demos/gifs/pingpong.gif">
At this point I’d suggest you to click that <a href="https://divan.dev/demos/pingpong/">link above</a> to the interactive WebGL animation (Ctrl/Cmd-Click to open it in a new tab) and play around with interactively. You can slowdown animation, speedup and see it in different angles.</p>
<p>Now, let’s run three players instead of two.</p>
<pre><code>    <span>go</span> player(table)
    <span>go</span> player(table)
    <span>go</span> player(table)
</code></pre>
<p><a href="https://divan.dev/demos/pingpong3/">Go to interactive WebGL animation</a>
<img alt="Ping-Pong 3" src="https://divan.dev/demos/gifs/pingpong3.gif">
We can see here that each player takes its turn sequentially and you may wonder why is it so. Why we see this strict order in goroutines receiving the ball?</p>
<p>The answer is because Go runtime holds waiting <a href="https://github.com/golang/go/blob/master/src/runtime/chan.go#L34">FIFO queue for receivers</a> (goroutines ready to receive on the particular channel), and in our case every player gets ready just after he passed the ball on the table. Let’s check it with more complex example and run 100 table tennis players.</p>
<pre><code><span>for</span> i := <span>0</span>; i &lt; <span>100</span>; i++ {
    <span>go</span> player(table)
}</code></pre>
<p><a href="https://divan.dev/demos/pingpong100/">Go to interactive WebGL animation</a>
<img alt="Ping-Pong 100" src="https://divan.dev/demos/gifs/pingpong100.gif">
The FIFO order is now obvious, isn’t it? We can spawn a million goroutines (they’re cheap), but for our goal that would be overkill.
Let’s see something different to play with. For example, common messaging patterns.</p>
<h3>Fan-In</h3>
<p>One of the popular patterns in the concurrent world is a so called <em>fan-in</em> pattern. It’s the opposite of the <em>fan-out</em> pattern, which we will cover later. To be short, fan-in is a function reading from the multiple inputs and multiplexing all into the single channel.</p>
<p>For example:</p>
<pre><code><span>package</span> main

<span>import</span> (
    <span>"fmt"</span>
    <span>"time"</span>
)

<span>func</span> producer(ch <span>chan</span> <span>int</span>, d time.Duration) {
    <span>var</span> i <span>int</span>
    <span>for</span> {
        ch &lt;- i
        i++
        time.Sleep(d)
    }
}

<span>func</span> reader(out <span>chan</span> <span>int</span>) {
    <span>for</span> x := <span>range</span> out {
        fmt.Println(x)
    }
}

<span>func</span> main() {
    ch := <span>make</span>(<span>chan</span> <span>int</span>)
    out := <span>make</span>(<span>chan</span> <span>int</span>)
    <span>go</span> producer(ch, <span>100</span>*time.Millisecond)
    <span>go</span> producer(ch, <span>250</span>*time.Millisecond)
    <span>go</span> reader(out)
    <span>for</span> i := <span>range</span> ch {
        out &lt;- i
    }
}</code></pre>
<p><a href="https://divan.dev/demos/fanin/">Go to interactive WebGL animation</a>
<img alt="Fan-In Pattern" src="https://divan.dev/demos/gifs/fanin.gif">
As we can see, first <em>producer</em> generates values each 100 milliseconds, and second one - each 250 milliseconds, but <em>reader</em> receives values from both producers immediately. Effectively, multiplexing happens in a range loop in <em>main</em>.</p>
<h3>Workers</h3>
<p>The opposite pattern to <em>fan-in</em> is a <em>fan-out</em> or <em>workers</em> pattern. Multiple goroutines can read from a single channel, distributing an amount of work between CPU cores, hence the <em>workers</em> name. In Go, this pattern is easy to implement - just start a number of goroutines with channel as parameter, and just send values to that channel - distributing and multiplexing will be done by Go runtime, automagically :)</p>
<pre><code><span>package</span> main

<span>import</span> (
    <span>"fmt"</span>
    <span>"sync"</span>
    <span>"time"</span>
)

<span>func</span> worker(tasksCh &lt;-<span>chan</span> <span>int</span>, wg *sync.WaitGroup) {
    <span>defer</span> wg.Done()
    <span>for</span> {
        task, ok := &lt;-tasksCh
        <span>if</span> !ok {
            <span>return</span>
        }
        d := time.Duration(task) * time.Millisecond
        time.Sleep(d)
        fmt.Println(<span>"processing task"</span>, task)
    }
}

<span>func</span> pool(wg *sync.WaitGroup, workers, tasks <span>int</span>) {
    tasksCh := <span>make</span>(<span>chan</span> <span>int</span>)

    <span>for</span> i := <span>0</span>; i &lt; workers; i++ {
        <span>go</span> worker(tasksCh, wg)
    }

    <span>for</span> i := <span>0</span>; i &lt; tasks; i++ {
        tasksCh &lt;- i
    }

    <span>close</span>(tasksCh)
}

<span>func</span> main() {
    <span>var</span> wg sync.WaitGroup
    wg.Add(<span>36</span>)
    <span>go</span> pool(&amp;wg, <span>36</span>, <span>50</span>)
    wg.Wait()
}</code></pre>
<p><a href="https://divan.dev/demos/workers/">Go to interactive WebGL animation</a>
<img alt="Workers" src="https://divan.dev/demos/gifs/workers.gif">
One thing worth to note here: the parallelism. As you can see, all goroutines ‘run’ in parallel, waiting for channel to give them ‘work’ to do. Given the animation above, it’s easy to spot that goroutines receive their work almost immediately one after another. Unfortunately, this animation doesn’t show in color where goroutine really does work or just waits for input, but this exact animation was recorded with GOMAXPROCS=4, so only 4 goroutines effectively run in parallel. We will get to this subject shortly.</p>
<p>For now, let’s do something more complex, and start workers that have their own workers (subworkers).</p>
<pre><code><span>package</span> main

<span>import</span> (
    <span>"fmt"</span>
    <span>"sync"</span>
    <span>"time"</span>
)

<span>const</span> (
    WORKERS    = <span>5</span>
    SUBWORKERS = <span>3</span>
    TASKS      = <span>20</span>
    SUBTASKS   = <span>10</span>
)

<span>func</span> subworker(subtasks <span>chan</span> <span>int</span>) {
    <span>for</span> {
        task, ok := &lt;-subtasks
        <span>if</span> !ok {
            <span>return</span>
        }
        time.Sleep(time.Duration(task) * time.Millisecond)
        fmt.Println(task)
    }
}

<span>func</span> worker(tasks &lt;-<span>chan</span> <span>int</span>, wg *sync.WaitGroup) {
    <span>defer</span> wg.Done()
    <span>for</span> {
        task, ok := &lt;-tasks
        <span>if</span> !ok {
            <span>return</span>
        }

        subtasks := <span>make</span>(<span>chan</span> <span>int</span>)
        <span>for</span> i := <span>0</span>; i &lt; SUBWORKERS; i++ {
            <span>go</span> subworker(subtasks)
        }
        <span>for</span> i := <span>0</span>; i &lt; SUBTASKS; i++ {
            task1 := task * i
            subtasks &lt;- task1
        }
        <span>close</span>(subtasks)
    }
}

<span>func</span> main() {
    <span>var</span> wg sync.WaitGroup
    wg.Add(WORKERS)
    tasks := <span>make</span>(<span>chan</span> <span>int</span>)

    <span>for</span> i := <span>0</span>; i &lt; WORKERS; i++ {
        <span>go</span> worker(tasks, &amp;wg)
    }

    <span>for</span> i := <span>0</span>; i &lt; TASKS; i++ {
        tasks &lt;- i
    }

    <span>close</span>(tasks)
    wg.Wait()
}</code></pre>
<p><a href="https://divan.dev/demos/workers2/">Go to interactive WebGL animation</a>
<img alt="Workers of workers" src="https://divan.dev/demos/gifs/workers2.gif">
Nice. Of course, we can set number of workers and subworkers to much higher values, but I tried to make animations clear and understandable.</p>
<p>There are even cooler fan-out patterns do exist, like the dynamic amount of workers/subworkers, with sending channels over channels, but the idea of fan-out should be clear for now.</p>
<h3>Servers</h3>
<p>Next common pattern is similar to fan-out, but with goroutines spawned for the short period of time, just to accomplish some task. It’s typically used for implementing servers - create a listener, run accept() in a loop and start goroutine for each accepted connection. It’s very expressive and allows to implement server handlers as simple as possible. Take a look at this simple example:</p>
<pre><code><span>package</span> main

<span>import</span> <span>"net"</span>

<span>func</span> handler(c net.Conn) {
    c.Write([]<span>byte</span>(<span>"ok"</span>))
    c.Close()
}

<span>func</span> main() {
    l, err := net.Listen(<span>"tcp"</span>, <span>":5000"</span>)
    <span>if</span> err != <span>nil</span> {
        <span>panic</span>(err)
    }
    <span>for</span> {
        c, err := l.Accept()
        <span>if</span> err != <span>nil</span> {
            <span>continue</span>
        }
        <span>go</span> handler(c)
    }
}</code></pre>
<p><a href="https://divan.dev/demos/servers/">Go to interactive WebGL animation</a>
<img alt="Servers" src="https://divan.dev/demos/gifs/servers.gif">
It’s not very interesting - it seems there is nothing happens in terms of concurrency. Of course, under the hood there is a ton of complexity, which is deliberately hidden from us. <a href="https://www.youtube.com/watch?v=rFejpH_tAHM">“Simplicity is complicated”</a>.</p>
<p>But let’s go back to concurrency and add some interaction to our server. Let’s say, each handler wants to write asynchronously to the logger. Logger itself, in our example, is a separate goroutine which does the job.</p>
<pre><code><span>package</span> main

<span>import</span> (
    <span>"fmt"</span>
    <span>"net"</span>
    <span>"time"</span>
)

<span>func</span> handler(c net.Conn, ch <span>chan</span> <span>string</span>) {
    ch &lt;- c.RemoteAddr().String()
    c.Write([]<span>byte</span>(<span>"ok"</span>))
    c.Close()
}

<span>func</span> logger(ch <span>chan</span> <span>string</span>) {
    <span>for</span> {
        fmt.Println(&lt;-ch)
    }
}

<span>func</span> server(l net.Listener, ch <span>chan</span> <span>string</span>) {
    <span>for</span> {
        c, err := l.Accept()
        <span>if</span> err != <span>nil</span> {
            <span>continue</span>
        }
        <span>go</span> handler(c, ch)
    }
}

<span>func</span> main() {
    l, err := net.Listen(<span>"tcp"</span>, <span>":5000"</span>)
    <span>if</span> err != <span>nil</span> {
        <span>panic</span>(err)
    }
    ch := <span>make</span>(<span>chan</span> <span>string</span>)
    <span>go</span> logger(ch)
    <span>go</span> server(l, ch)
    time.Sleep(<span>10</span> * time.Second)
}</code></pre>
<p><a href="https://divan.dev/demos/servers2/">Go to interactive WebGL animation</a>
<img alt="Servers 2" src="https://divan.dev/demos/gifs/servers2.gif">
Quite demonstrative, isn’t it? But it’s easy to see that our <em>logger</em> goroutine can quickly become a bottleneck if the number of requests increase and logging action take some time (preparing and encoding data, for example). We can use an already known fan-out pattern. Let’s do it.</p>
<h3>Server + Worker</h3>
<p>Server with worker example is a bit advanced version of the logger. It not only does some work but sends the result of its work back to the pool using <em>results</em> channel. Not a big deal, but it extends our logger example to something more practical.</p>
<p>Let’s see the code and animation:</p>
<pre><code><span>package</span> main

<span>import</span> (
    <span>"net"</span>
    <span>"time"</span>
)

<span>func</span> handler(c net.Conn, ch <span>chan</span> <span>string</span>) {
    addr := c.RemoteAddr().String()
    ch &lt;- addr
    time.Sleep(<span>100</span> * time.Millisecond)
    c.Write([]<span>byte</span>(<span>"ok"</span>))
    c.Close()
}

<span>func</span> logger(wch <span>chan</span> <span>int</span>, results <span>chan</span> <span>int</span>) {
    <span>for</span> {
        data := &lt;-wch
        data++
        results &lt;- data
    }
}

<span>func</span> parse(results <span>chan</span> <span>int</span>) {
    <span>for</span> {
        &lt;-results
    }
}

<span>func</span> pool(ch <span>chan</span> <span>string</span>, n <span>int</span>) {
    wch := <span>make</span>(<span>chan</span> <span>int</span>)
    results := <span>make</span>(<span>chan</span> <span>int</span>)
    <span>for</span> i := <span>0</span>; i &lt; n; i++ {
        <span>go</span> logger(wch, results)
    }
    <span>go</span> parse(results)
    <span>for</span> {
        addr := &lt;-ch
        l := <span>len</span>(addr)
        wch &lt;- l
    }
}

<span>func</span> server(l net.Listener, ch <span>chan</span> <span>string</span>) {
    <span>for</span> {
        c, err := l.Accept()
        <span>if</span> err != <span>nil</span> {
            <span>continue</span>
        }
        <span>go</span> handler(c, ch)
    }
}

<span>func</span> main() {
    l, err := net.Listen(<span>"tcp"</span>, <span>":5000"</span>)
    <span>if</span> err != <span>nil</span> {
        <span>panic</span>(err)
    }
    ch := <span>make</span>(<span>chan</span> <span>string</span>)
    <span>go</span> pool(ch, <span>4</span>)
    <span>go</span> server(l, ch)
    time.Sleep(<span>10</span> * time.Second)
}</code></pre>
<p><a href="https://divan.dev/demos/servers3/">Go to interactive WebGL animation</a>
<img alt="Server + Worker" src="https://divan.dev/demos/gifs/servers3.gif">
We distributed work between 4 goroutines, effectively improving the throughput of the logger, but from this animation, we can see that logger still may be the source of problems. Thousands of connections converge in a single channel before being distributed and it may result in a logger being bottleneck again. But, of course, it will happen on much higher loads.</p>
<h3>Concurrent Prime Sieve</h3>
<p>Enough fan-in/fan-out fun. Let’s see more sophisticated concurrency algorithms. One of my favorite examples is a Concurrent Prime Sieve, found in <a href="https://talks.golang.org/2012/concurrency.slide">“Go Concurrency Patterns”</a> talk. Prime Sieve, or <a href="https://en.wikipedia.org/wiki/Sieve_of_Eratosthenes">Sieve of Eratosthenes</a> is an ancient algorithm for finding prime number up to the given limit. It works by eliminating multiples of all primes in a sequential manner. Naive algorithm is not really efficient, especially on multicore machines.</p>
<p>The concurrent variant of this algorithm uses goroutines for filtering numbers - one goroutine per every prime discovered, and channels for sending numbers from the generator to filters. When prime is found, it’s being sent via the channel to the <em>main</em> for output. Of course, this algorithm is also not very efficient, especially if you want to find large primes and look for the lowest Big O complexity, but I find it extremely elegant.</p>
<pre><code>
<span>package</span> main

<span>import</span> <span>"fmt"</span>

<span>func</span> Generate(ch <span>chan</span>&lt;- <span>int</span>) {
    <span>for</span> i := <span>2</span>; ; i++ {
        ch &lt;- i 
    }
}

<span>func</span> Filter(in &lt;-<span>chan</span> <span>int</span>, out <span>chan</span>&lt;- <span>int</span>, prime <span>int</span>) {
    <span>for</span> {
        i := &lt;-in 
        <span>if</span> i%prime != <span>0</span> {
            out &lt;- i 
        }
    }
}

<span>func</span> main() {
    ch := <span>make</span>(<span>chan</span> <span>int</span>) 
    <span>go</span> Generate(ch)      
    <span>for</span> i := <span>0</span>; i &lt; <span>10</span>; i++ {
        prime := &lt;-ch
        fmt.Println(prime)
        ch1 := <span>make</span>(<span>chan</span> <span>int</span>)
        <span>go</span> Filter(ch, ch1, prime)
        ch = ch1
    }
}</code></pre>
<p><a href="https://divan.dev/demos/primesieve/">Go to interactive WebGL animation</a>
<img alt="PrimeSieve" src="https://divan.dev/demos/gifs/primesieve.gif">
Feel free to play with this animation in interactive mode. I like how illustrative it is - it really can help understand this algorithm better. The <em>generate</em> goroutine emits every integer number, starting from 2, and each new goroutine filters out only specific prime multiples - 2, 3, 5, 7…, sending first found prime to <em>main</em>. If you rotate it to see from the top, you’ll see all numbers being sent from goroutines to main are prime numbers. Beautiful algorithm, especially in 3D.</p>
<h3>GOMAXPROCS</h3>
<p>Now, let’s go back to our workers example. Remember, I told that it was run with GOMAXPROCS=4? That’s because all these animations are not art work, they are real traces of real programs.</p>
<p>Let’s refresh our memory on what <a href="https://golang.org/pkg/runtime/#GOMAXPROCS">GOMAXPROCS</a> is.</p>
<blockquote>
<p>GOMAXPROCS sets the maximum number of CPUs that can be executing simultaneously.</p>
</blockquote>
<p>CPU means logical CPU, of course. I modified workers example a bit to make a real work (not just sleep) and use real CPU time. Then I ran the code without any modification except setting different GOMAXPROCS value. The Linux box had 2 CPUs with 12 cores each, resulting in 24 cores.</p>
<p>So, the first run demonstrates the program running on 1 core, and second - using the power of all 24 cores availiable.</p>
<p><a href="https://divan.dev/demos/gomaxprocs1/">WebGL animation - 1</a> | <a href="https://divan.dev/demos/gomaxprocs24/">WebGL animation - 24</a>
<img alt="GOMAXPROCS1" src="https://divan.dev/demos/gifs/gomaxprocs1.gif"> <img alt="GOMAXPROCS24" src="https://divan.dev/demos/gifs/gomaxprocs24.gif">
The time speed in these animations are different (I wanted all animations to fit the same time/height), so the difference is obvious. With GOMAXPROCS=1, next worker start real work only after previous has finish it’s work. With GOMAXPROCS=24 the speedup is huge, and overhead for multiplexing is negligible.</p>
<p>It’s important to understand, though, that increasing GOMAXPROCS not always boosts performance, and there cases when it actually makes it worse.</p>
<h3>Goroutines leak</h3>
<p>What else we can demonstrate from concurrent things in Go? The one thing that comes to my mind is a goroutines leak. A leak can happen, for example, if you <a href="http://openmymind.net/Leaking-Goroutines/">start goroutine but it falls out of scope</a>. Or you simply forget to add finish condition, and run a for {} loop.</p>
<p>First time I’ve encountered goroutine leak in my code, the scary image appeared in my head and I wrote <a href="https://github.com/divan/expvarmon">expvarmon</a> next weekend. And now I can visualize that scary image with WebGL.</p>
<p>Take a look:</p>
<p><a href="https://divan.dev/demos/leak/">Go to interactive WebGL animation</a>
<img alt="Leak" src="https://divan.dev/demos/gifs/leak.gif">
I feel pain even simply by looking at this :) All those lines are wasted resources and a ticking bomb for your program.</p>
<h3>Parallelism is not Concurrency</h3>
<p>The last thing I want to illustrate is a difference between parallelism and concurrency. This topic is <a href="https://existentialtype.wordpress.com/2011/03/17/parallelism-is-not-concurrency/">well</a> <a href="https://ghcmutterings.wordpress.com/2009/10/06/parallelism-concurrency/">covered</a>, and there is a <a href="https://www.youtube.com/watch?v=cN_DpYBzKso">great talk by Rob Pike</a> on the subject. One of the #mustwatch videos, really.</p>
<p>
</p>
<center>

</center>

<p>To be short,</p>
<blockquote>
<p>Parallelism is simply running things in parallel.</p>
<p>Concurrency is a way to structure your program.</p>
</blockquote>
<p>Thus, the concurrent program may or may not be parallel, these concepts are somehow orthogonal. We have seen this earlier in a demonstration of GOMAXPROCS setting effects.</p>
<p>I can repeat all those linked articles and talks, but a picture is worth a thousand words. What I can do here is to visualize the difference.
So, this is parallelism. Many things running in parallel.</p>
<p><a href="https://divan.dev/demos/parallelism1/">Go to interactive WebGL animation</a>
<img alt="Parallelism 1" src="https://divan.dev/demos/gifs/parallelism1.gif"></p>
<p>And this is also parallelism:</p>
<p><a href="https://divan.dev/demos/parallelism2/">Go to interactive WebGL animation</a>
<img alt="Parallelism 2" src="https://divan.dev/demos/gifs/parallelism2.gif"></p>
<p>But this is concurrency:</p>
<p><img alt="PrimeSieve" src="https://divan.dev/demos/gifs/primesieve.gif"></p>
<p>And this:</p>
<p><img alt="Workers of workers" src="https://divan.dev/demos/gifs/workers2.gif"></p>
<p>And this is also concurrency:</p>
<p><img alt="Ping-Pong 100" src="https://divan.dev/demos/gifs/pingpong100.gif"></p>
<h3>How it was made</h3>
<p>To create these animations, I wrote two programs: <em>gotracer</em> and <em>gothree.js</em> library.
First, gotracer does the following:</p>
<ul>
<li>parse AST tree of Go program and insert special commands with output on concurrency related events - start/stop goroutine, create a channel, send/receive to/from a channel.</li>
<li>run generated program</li>
<li>analyze this special output and produce JSON with the description of events and timestamps.</li>
</ul>
<p>Example of the resulting JSON:
<img alt="JSON sample" src="https://divan.dev/images/sshot_json.png"></p>
<p>Next, gothree.js uses the power of an amazing <a href="http://threejs.org/">Three.js</a> library to draw 3D lines and objects using WebGL. Little wrapper to fit into single html page - and there it is.</p>
<p>This approach, though, is super limited. I have to accurately choose examples, rename channels and goroutines to make more or less complex code to produce a correct trace. With this approach, there is no easy way to correllate channels between goroutines if they have different names. Not to mention channels sent over channels of type chan. There are also huge issues with timing - output to stdout can take more time than sending value, so in some cases I had to place time.Sleep(some amount of milliseconds) to get proper animation.</p>
<p>Basically, that is a reason why I’m not open-sourcing the code yet. I’m playing with Dmitry Vyukov’s <a href="https://golang.org/cmd/trace/">execution tracer</a>, it seems to provide good level of details of events, but do not contain info on which values are being sent. Maybe there are better ways to achieve the desired goal. Write me in twiter or in comments here, if you have ideas. It would be super great to extend this two-weekends tool to be a real debugging/tracing instrument suitable for any Go program.</p>
<p>I also would be happy to visualize more interesting concurrent algorithms and patterns not listed here. Feel free to write ones in the comments.</p>
<p>Happy coding!</p>
<p>UPD: This tool is available at <a href="https://github.com/divan/gotrace">github.com/divan/gotrace</a> and using Go Execution Tracer and patched runtime to generate trace.</p>
<p>Also, I’m open for the new job, so if you’re interesting company/team, have challenging problems to solve, use Go, open for remote (or you’re in Barcelona) and hiring, let me know :)</p>
</section></div><br>
<br>
<br>
<br>
<br>
<br>
via divan's blog https://divan.dev<br>
<br>
March 14, 2019 at 04:47PM
